src/lib.rs

```diff
@@ -474,6 +493,156 @@ impl XenControl {
         last_error!(self, max_gpfn)
     }
 
+    pub fn vcpu_getinfo(&self, domid: u32, vcpu: u32) -> Result<xc_vcpuinfo_t, XcError> {
+        debug!("vcpu_getinfo");
+        let xc = self.handle.as_ptr();
+        let mut vcpu_info = unsafe { mem::zeroed() };
+
+        (self.libxenctrl.clear_last_error)(xc);
+        (self.libxenctrl.vcpu_getinfo)(xc, domid, vcpu, &mut vcpu_info);
+
+        last_error!(self, vcpu_info)
+    }
+
+    pub fn physinfo(&self) -> Result<xc_physinfo_t, XcError> {
+        debug!("physinfo");
+        let xc = self.handle.as_ptr();
+        let mut physinfo = unsafe { mem::zeroed() };
+
+        (self.libxenctrl.clear_last_error)(xc);
+        (self.libxenctrl.physinfo)(xc, &mut physinfo);
+
+        last_error!(self, physinfo)
+    }
+
+    pub fn get_cpuinfo<'a>(
+        &'_ self,
+        infos: &'a mut [mem::MaybeUninit<xc_cpuinfo_t>],
```

Comment:
As an API user, I believe this will force me to create an array of `xc_cpuinfo_t`, which means i probably need to use `std::mem::zeroed`, and therefore introduce some unsafe code in my app (which defeats the purpose of a safe wrapper on top of `xenctrl-sys` ) :)

It's a mistake I made when I quickly added the functions to listen for VMI events for libmicrovmi.
Example in the Xen driver when putting the response back on the ring buffer:
https://github.com/Wenzel/libmicrovmi/blob/master/src/driver/xen.rs#L446

I think we should wrap this type as well.
Plus we could update our parameter type to be more generic, and accept either an array or a vector, with `AsRef`


What do you think ?

Reply:
In this case, `infos` is actually a `MaybeUninit` slice (which doesn't require unsafe as long as we don't try to access the undelying structure) that we can create using e.g `MaybeUninit::zeroed` (https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#method.zeroed which is a safe function).

The idea is that the user provide a buffer (`infos`) and the function returns a part of this buffer that will be initialized with data and slice of it is returned (only initialized part) directly usable without having to rely on unsafe code.

I am not sure on another approach as :
- cpu count is not well known (but `physinfo` can provide it with `nr_cpus` field)
- we need to rely on allocations on each function call to create the cpuinfo buffer, which isn't really that bad, but not the best either

Reply:
Rethinking about this, do we need this API to be as efficient as possible (like `get_cxstat`) ?

Otherwise, we can change the definition to returning a `Result<>`:

```rust
pub fn get_cpuinfo(&self, max_cpus: i32) -> Result<Vec<xc_cpuinfo_t>, XcError> {
  // 
}
```

And if the Vec heap allocation is becoming problematic, we can try [`smallvec`](https://docs.rs/smallvec/latest/smallvec/) which is a stack allocated Vec that falls back to the heap if the allocated size becomes too large.

Reply:
Implementation could look like this, and is much simpler:

```rust
    pub fn get_cpuinfo(&self, max_cpus: i32) -> Result<Vec<xc_cpuinfo_t>, XcError> {
        debug!("get_cpuinfo");
        let mut infos = vec![xc_cpuinfo_t::default(); max_cpus.try_into().unwrap()];
        let mut nr_cpus: i32 = 0;

        let rc = (self.libxenctrl.get_cpuinfo)(
            self.handle.as_ptr(),
            infos.len() as i32,
            infos.as_mut_ptr() as _,
            &mut nr_cpus,
        );

        infos.truncate(nr_cpus);

        last_error!(self, infos, rc)
    }
``